### ISSUE TITLE
Async style with reply.send

### ISSUE THREAD

**Issue by StarpTech** [fastify/fastify#1864]
## 🐛 Bug Report

`repl.send` inside an async function has created a lot of headache.

- https://github.com/fastify/fastify/issues/238
- https://github.com/fastify/fastify/issues/238
- https://github.com/fastify/fastify/issues/548
- https://github.com/fastify/fastify/issues/1131

with https://github.com/fastify/fastify-static/issues/110 we have another issue. The issue ist that an error is logged because `sendFile` doesn't accept a callback or return a promise. The route is fulfilled to `undefined` before the file was sent. This behaviour doesn't agree with https://github.com/fastify/fastify/blob/master/docs/Routes.md#promise-resolution.

Due to the current implementation every decorator/plugin which calls `.send` under the hood must provide a callback or promise to the consumer. If they don't the handler can't use async/await and will loose the control over the response.

## Expected behavior

**In the longer term we change the api.**
```js
// the data flow isnt determined by the function return value (koajs be like)
fastify
  .get('/', async (request, reply) => {
    reply.body = await data()
	repl.body = sync()
  });
```
**In short term:**
Teach and update the docs that you can use only `reply.send` in async/await if you still return a promise that controls the asynchronous operation.
```js
until send is called
fastify
  .get('/', async (request, reply) => {
    return op().then(reply.send)
  });
```
or we recommend to not use `rely.send` at all in async functions

**Comment by jsumners** [comment]
I am a strong 👎 on removing support for:

```js
fastify.get('/', async function (req, reply) {
  // do some promise stuff
  // ...
  reply.send({hello: 'world'})
})
```

The only reason I use `async` handlers is to work with promise based functions before issuing the reply.

**Comment by StarpTech** [comment]
@jsumners if you use only promises you don't need `async/await`

```js
fastify.get('/', function (req, reply) {
  // do some promise stuff
  // ...
  return Promise.resolve()
})
```

if you use `async` you should use the corresponding `await` style
```js
fastify.get('/', async function (req, reply) {
  // do some await stuff
  // ...
  return value;
})
```

if you use `callback-style` you can just use `.send`
```js
fastify.get('/', function (req, reply) {
  // do some callback stuff
  // ...
  reply.send({hello: 'world'})
})
```

but as soon as you try to mix different styles it gets hairy very fast and this is what I'm talking about. We recommend/support different styles.

**Comment by jsumners** [comment]
The async operations have nothing to do with _sending_ the response, only building the response. I’m not listing a complete example here because it is beside the point. Which is: `reply.send` should work regardless of the function declaration.

**Comment by StarpTech** [comment]
@jsumners removing support for `reply.send` is not my intention. I'd like to start a discussion because a combination of callback and async/await has lead to confusions in the community.

> The async operations have nothing to do with sending the response, only building the response. I’m not listing a complete example here because it is beside the point. Which is: reply.send should work regardless of the function declaration.

I don't know if I understood you correctly but with the use of `async/await` or `promises` the control flow should be stay in the promise chain not by a callback. That's how it should work.

**Comment by StarpTech** [comment]
The best example how it should work is Hapi.js :smiley_cat: They always returned a `Response` object which is build by the request `tooklit` OR a javascript value.

**Comment by jsumners** [comment]
```js
async function handler (req, reply) {
  const someResult = await anAsyncFunction(req.body)
  if (someResult.error) {
    return reply.code(500).send({error: someResult.error})
  }

  reply.send({success: 'is-you'})
}
```

That's the simplest I can make it. *There are use cases* for using `reply.send` in an `async` function. Simply returning the resolution, or rejection, of a promise *is not sufficient* in all cases.

**Comment by StarpTech** [comment]
That's no real use case it's your favorite way to write code. 

You can just return the value and in case of error you can return the value with an rejected promise.

Hapi removed callback style in route handlers and Hapi is driven by a big community and enterprise users and this significant.

**Comment by jsumners** [comment]
Amazing. I can't even begins to respond in a cordial manner to this.

**Comment by StarpTech** [comment]
I don't want to be rude. I just want to convince you. You talked about use cases which can be handled only with reply.send which ones?

**Comment by mcollina** [comment]
I think using `reply.send()` within async functions is a pretty solid case, however I agree that the developer experience it's currently not there yet.

How about we support:

```js
fastify
  .get('/', async (request, reply) => {
    someAsync(() => reply.send())
    await reply
  });
```

This would be possible by making reply a thenable. This would be possible just by implenting [`then`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then)

**Comment by StarpTech** [comment]
@mcollina what's the motivation of using `reply.send()` in an async function? Just to allow mixing different styles? I don't get it :confused:

Only to ensure: I'm talking about the future fastify v3-4 not to patch the current codebase. I also created a [PR](https://github.com/fastify/fastify/pull/1868) to improve the documentation around Promise/async-await usage.

**Comment by mcollina** [comment]
> @mcollina what's the motivation of using reply.send() in an async function? Just to allow mixing different styles? I don't get it 😕

When interacting with other library it's almost impossible to avoid. The proposed syntax will not work with the `sendFile` issue for example, as we will have to wait anyway for `send` to finish. By making `reply` awaitable, we can provide an escape hatch for all those cases without making any significant change to our APIs.

**Comment by delvedor** [comment]
I agree with @mcollina, there are some cases where you need to use `reply.send`, and Fastify is designed to allow you to use both callbacks and promises style.
Another case that we will fix with this change, is to enable an easier support for async style handlers with [`point-of-view`](https://github.com/fastify/point-of-view).
Making `reply` a thenable might bot be the perfect solution, but it solves in a smart way the problem.

**Comment by StarpTech** [comment]
@mcollina OK you are talking about scenarios to avoid promisify things right? I'm fine with that proposal even when it makes the api more complicated.

**Comment by mcollina** [comment]
> @mcollina OK you are talking about scenarios to avoid promisify things right? I'm fine with that proposal even when it makes the api more complicated.

Yes. This is both the case of `point-of-view` and `fastify-static`. Promisifying is often an all-or-nothing choice, and there are plenty of cases where it's impractical.

**Comment by jsumners** [comment]
Thank you @mcollina for stepping in here. I appreciate a level head recognizing the problem and offering up an excellent solution.

**Comment by mcollina** [comment]
Thanks!

**Comment by mkhoussid** [comment]
A bit late to the party and maybe not posting in the correct place, but I came across this situation today. My endpoint looks kind of like this,

```javascript
async function getOtp(
	request: FastifyRequestTypebox<typeof LoginSchema>,
	reply: FastifyReplyTypebox<typeof LoginSchema>,
) => {
        // do async/await validation, db queries
        reply.send();
	await sendOtpEmail({ id: 1234 });
},
```
The email client I'm using takes on average about 4-5 seconds to resolve successfully. It's not terribly long, but I've decided to just send the email after calling `reply.send`, so that the user doesn't see a full screen spinner during the time they wait.

Anyway, I presume that this isn't possible to do?

**Comment by Uzlopak** [comment]
Actually... to implement something like this, I would not immediatly send the email from the handler. 

I would actually create a "Queue", where you put the otpEmails into the queue. The processor the queue tries to send the email, and if it fails retries. By decoupling this, you can solve this imho better.

**Comment by mateus4k** [comment]
@mkhoussid, I solved my similar problem by doing this:

```ts
async function getOtp(request, reply) {
  await reply.send();
  await sendOtpEmail({ id: 1234 });
}
```
